airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -0,0 +1,465 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination.bigquery.typing_deduping;
+
+import static java.util.stream.Collectors.joining;
+
+import com.google.cloud.bigquery.StandardSQLTypeName;
+import com.google.cloud.bigquery.StandardTableDefinition;
+import com.google.cloud.bigquery.TableDefinition;
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.AirbyteProtocolType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.Array;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.OneOf;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.Struct;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.UnsupportedOneOf;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteTypeUtils;
+import io.airbyte.integrations.base.destination.typing_deduping.CatalogParser.StreamConfig;
+import io.airbyte.integrations.base.destination.typing_deduping.SqlGenerator;
+import io.airbyte.integrations.destination.bigquery.BigQuerySQLNameTransformer;
+import io.airbyte.protocol.models.v0.DestinationSyncMode;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.text.StringSubstitutor;
+
+public class BigQuerySqlGenerator implements SqlGenerator<TableDefinition> {
+
+  public static final String QUOTE = "`";
+  private static final BigQuerySQLNameTransformer nameTransformer = new BigQuerySQLNameTransformer();
+
+  private final ColumnId CDC_DELETED_AT_COLUMN = buildColumnId("_ab_cdc_deleted_at");
+
+  @Override
+  public StreamId buildStreamId(final String namespace, final String name, final String rawNamespaceOverride) {
+    return new StreamId(
+        // TODO is this correct?
+        nameTransformer.getNamespace(namespace),
+        nameTransformer.convertStreamName(name),
+        nameTransformer.getNamespace(rawNamespaceOverride),
+        nameTransformer.convertStreamName(namespace + "_" + name),
+        namespace,
+        name);
+  }
+
+  @Override
+  public ColumnId buildColumnId(final String name) {
+    String quotedName = name;
+
+    // Bigquery columns are case insensitive, so do all our validation on the lowercased name
+    String canonicalized = name.toLowerCase();
+
+    // Column names aren't allowed to start with certain strings. Prepend an underscore if this happens.
+    if (canonicalized.startsWith("_table_")
```

Comment:
```suggestion
    final List<String> invalidColumnPrefixes = List.of(
      "_table_",
      "_file_",
      "_partition_",
      "_row_timestamp_",
      "__root__",
      "_colidentifier_"
    );
    if invalidColumnPrefixes.stream().anyMatch(col -> canonicalized.startsWith(col)) {
      quotedName = "_" + quotedName;
      canonicalized = "_" + canonicalized;
    }
```

Reply:
this forces a tmp variable for `canonicalized` b/c lambdas can only use (effectively) final variables
```java
    final String wipCanonicalized = name.toLowerCase();

    // Column names aren't allowed to start with certain strings. Prepend an underscore if this happens.
    final List<String> invalidColumnPrefixes = List.of(
        "_table_",
        "_file_",
        "_partition_",
        "_row_timestamp_",
        "__root__",
        "_colidentifier_"
    );
    final String canonicalized;
    if (invalidColumnPrefixes.stream().anyMatch(wipCanonicalized::startsWith)) {
      quotedName = "_" + quotedName;
      canonicalized = "_" + wipCanonicalized;
    } else {
      canonicalized = wipCanonicalized;
    }
```

I very slightly prefer the original, but happy to make the change if you prefer this version

Reply:
Thats what I get for trying to write Java without an IDE 😄 , you can get around it without a temporary variable by using the initial `name` argument, so 
```
if invalidColumnPrefixes.stream().anyMatch(col -> name.toLowerCase().startsWith(col)) {
```
and I prefer this approach because it keeps the conditional logic separate from the values its checking against, and is just a little simpler to read since its always a `startsWith` check. Think its a good DRY example because it would be easy enough to mistype `canonicalized` with one of the other string values or something, or add an `endsWith` instead, etc.

Reply:
fair point - done in https://github.com/airbytehq/airbyte/pull/28057/commits/aabdcbdb9842f516e6a76cba5290d1a146476e8b